File: src/analyzers/secrets.py
from __future__ import annotations

from pathlib import Path

from src.models import Finding, Location, RepoScanResult
from src.utils.normalize import is_non_empty_placeholder
from src.utils.ordering import sort_findings


def secret_and_placeholder_findings(scan_result: RepoScanResult) -> tuple[Finding, ...]:
    findings: list[Finding] = []
    seen: set[tuple[str, str, int, str]] = set()

    for definition in scan_result.definitions:
        if Path(definition.file_path).name != ".env":
            continue

        if definition.normalized_value_kind == "secret_like":
            key = ("ENV008", definition.file_path, definition.line_number, definition.name)
            if key not in seen:
                seen.add(key)
                findings.append(
                    Finding(
                        code="ENV008",
                        severity="warning",
                        title="Secret-like committed value",
                        details=(
                            f"{definition.name} in {definition.file_path} looks like a real "
                            "secret value."
                        ),
                        variable_name=definition.name,
                        locations=(
                            Location(
                                file_path=definition.file_path,
                                line_number=definition.line_number,
                            ),
                        ),
                        confidence="low",
                        source_kind="heuristic",
                        reason="Value shape is long, opaque, and mixed-character.",
                        suppression_key=(
                            f"secret-like:{definition.file_path}:{definition.line_number}:{definition.name}"
                        ),
                    )
                )

        if is_non_empty_placeholder(definition.value):
            key = ("ENV009", definition.file_path, definition.line_number, definition.name)
            if key not in seen:
                seen.add(key)
                findings.append(
                    Finding(
                        code="ENV009",
                        severity="warning",
                        title="Placeholder-like committed value",
                        details=(
                            f"{definition.name} in {definition.file_path} uses a placeholder "
                            "value."
                        ),
                        variable_name=definition.name,
                        locations=(
                            Location(
                                file_path=definition.file_path,
                                line_number=definition.line_number,
                            ),
                        ),
                        confidence="low",
                        source_kind="heuristic",
                        reason=(
                            "Common placeholder values in committed .env files "
                            "are easy to miss."
                        ),
                        suppression_key=(
                            f"placeholder:{definition.file_path}:{definition.line_number}:{definition.name}"
                        ),
                    )
                )

    return sort_findings(findings)

File: internal/analyzers/secrets.go
package analyzers

import (
	"fmt"
	"path/filepath"

	"github.com/davisbuilds/envdiff/internal/model"
	"github.com/davisbuilds/envdiff/internal/normalize"
	"github.com/davisbuilds/envdiff/internal/order"
)

func SecretAndPlaceholderFindings(scanResult model.RepoScanResult) []model.Finding {
	findings := []model.Finding{}
	seen := map[string]struct{}{}

	for _, definition := range scanResult.Definitions {
		if filepath.Base(definition.FilePath) != ".env" {
			continue
		}

		if definition.NormalizedValueKind == "secret_like" {
			key := fmt.Sprintf("ENV008:%s:%d:%s", definition.FilePath, definition.LineNumber, definition.Name)
			if _, ok := seen[key]; !ok {
				seen[key] = struct{}{}
				findings = append(findings, model.Finding{
					Code:       "ENV008",
					Confidence: model.StringPtr("low"),
					Details: fmt.Sprintf(
						"%s in %s looks like a real secret value.",
						definition.Name,
						definition.FilePath,
					),
					Locations: []model.Location{{
						FilePath:   definition.FilePath,
						LineNumber: model.IntPtr(definition.LineNumber),
					}},
					Reason:     model.StringPtr("Value shape is long, opaque, and mixed-character."),
					Severity:   "warning",
					SourceKind: "heuristic",
					SuppressionKey: model.StringPtr(
						fmt.Sprintf(
							"secret-like:%s:%d:%s",
							definition.FilePath,
							definition.LineNumber,
							definition.Name,
						),
					),
					Title:        "Secret-like committed value",
					VariableName: model.StringPtr(definition.Name),
				})
			}
		}

		if normalize.IsNonEmptyPlaceholder(definition.Value) {
			key := fmt.Sprintf("ENV009:%s:%d:%s", definition.FilePath, definition.LineNumber, definition.Name)
			if _, ok := seen[key]; !ok {
				seen[key] = struct{}{}
				findings = append(findings, model.Finding{
					Code:       "ENV009",
					Confidence: model.StringPtr("low"),
					Details: fmt.Sprintf(
						"%s in %s uses a placeholder value.",
						definition.Name,
						definition.FilePath,
					),
					Locations: []model.Location{{
						FilePath:   definition.FilePath,
						LineNumber: model.IntPtr(definition.LineNumber),
					}},
					Reason: model.StringPtr(
						"Common placeholder values in committed .env files are easy to miss.",
					),
					Severity:   "warning",
					SourceKind: "heuristic",
					SuppressionKey: model.StringPtr(
						fmt.Sprintf(
							"placeholder:%s:%d:%s",
							definition.FilePath,
							definition.LineNumber,
							definition.Name,
						),
					),
					Title:        "Placeholder-like committed value",
					VariableName: model.StringPtr(definition.Name),
				})
			}
		}
	}

	return order.Findings(findings)
}
